unit_tests/MacroTranslTests.d

```diff
@@ -0,0 +1,566 @@
+/**
+ * Copyright: Copyright (c) 2016 Wojciech Szęszoł. All rights reserved.
+ * Authors: Wojciech Szęszoł
+ * Version: Initial created: Jun 02, 2016
+ * License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0)
+ */
+import core.exception;
+
+import Common;
+import Assert;
+
+import clang.c.Index;
+
+import dstep.translator.Context;
+import dstep.translator.MacroDefinition;
+import dstep.translator.Options;
+import dstep.translator.Output;
+
+private alias assertTMD = assertTranslatesMacroDefinition;
+private alias assertTME = assertTranslatesMacroExpression;
+private alias assertDPME = assertDoesntParseMacroExpression;
+
+// Translate basic macro definitions.
+unittest
+{
+    assertTMD(q"C
+#define FOO
+C", q"D
+D");
+
+    assertTMD(q"C
+#define FOO 1
+C", q"D
+enum FOO = 1;
+D");
+
+    assertTMD(q"C
+#define FOO "bar"
+C", q"D
+enum FOO = "bar";
+D");
+
+    assertTMD(q"C
+#define FOO() 0
+C", q"D
+extern (D) int FOO()
+{
+    return 0;
+}
+D");
+
+    assertTMD(q"C
+#define FOO(a, b) a + b
+C", q"D
+extern (D) auto FOO(T0, T1)(auto ref T0 a, auto ref T1 b)
+{
+    return a + b;
+}
+D");
+
+    assertTMD(q"C
+#define FOO() 0 + 1
+C", q"D
+extern (D) int FOO()
+{
+    return 0 + 1;
+}
+D");
+
+    assertTMD(q"C
+#define FOO(a, b, c) (\
+    ((a) * 100) \
+  + ((b) * 10) \
+  + c)
+C", q"D
+extern (D) auto FOO(T0, T1, T2)(auto ref T0 a, auto ref T1 b, auto ref T2 c)
+{
+    return (a * 100) + (b * 10) + c;
+}
+D");
+
+    assertTMD(q"C
+#define STRINGIZE(major, minor) #major"."#minor
+C", q"D
+extern (D) string STRINGIZE(T0, T1)(auto ref T0 major, auto ref T1 minor)
+{
+    import std.conv : to;
+
+    return to!string(major) ~ "." ~ to!string(minor);
+}
+D");
+
+    assertTMD(q"C
+#define VERSION ENCODE( \
+    MAJOR, \
+    MINOR)
+C", q"D
+enum VERSION = ENCODE(MAJOR, MINOR);
+D");
+
+}
+
+// Translate simple expressions.
+unittest
+{
+    assertTMD(q"C
+#define FOO(a) a + 1
+C", q"D
+extern (D) auto FOO(T)(auto ref T a)
+{
+    return a + 1;
+}
+D");
+
+    assertTMD(q"C
+#define FOO(a) a * 1 / 2 % 3
+C", q"D
+extern (D) auto FOO(T)(auto ref T a)
+{
+    return a * 1 / 2 % 3;
+}
+D");
+
+    assertTMD(q"C
+#define FOO(a) a -1 + 2
+C", q"D
+extern (D) auto FOO(T)(auto ref T a)
+{
+    return a - 1 + 2;
+}
+D");
+
+    assertTMD(q"C
+#define FOO(a) a << 1 >> 2
+C", q"D
+extern (D) auto FOO(T)(auto ref T a)
+{
+    return a << 1 >> 2;
+}
+D");
+
+    assertTMD(q"C
+#define FOO(a) a < 1 && a < 2 && 1 <= 2 && 2 >= 3 || 1 == 1 || a != a
+C", q"D
+extern (D) auto FOO(T)(auto ref T a)
+{
+    return a < 1 && a < 2 && 1 <= 2 && 2 >= 3 || 1 == 1 || a != a;
+}
+D");
+
+    assertTMD(q"C
+#define FOO(a) ((a) == 1 ? 0 : 1 ? 2 : 3)
+C", q"D
+extern (D) int FOO(T)(auto ref T a)
+{
+    return a == 1 ? 0 : 1 ? 2 : 3;
+}
+D");
+
+}
+
+// Translate cast operator (FIXME: type inference)
```

Comment:
It is about trying to guess the return type of macro. But I suppose it is unimportant now, as D's `auto` is good enough. I removed the fixme.